File: src/services/history.ts
import { invoke } from "@tauri-apps/api/core";

import type {
  ConversationDetail,
  ConversationSummary,
  HistoryBootstrap,
} from "@/types";

export const historyBootstrap = () =>
  invoke<HistoryBootstrap>("history_bootstrap");

export const historyListConversations = () =>
  invoke<ConversationSummary[]>("history_list_conversations");

export const historyGetConversation = (conversationId: string) =>
  invoke<ConversationDetail>("history_get_conversation", { conversationId });

export const historyGetConversationPage = (
  conversationId: string,
  beforeSeq?: number | null,
  limit?: number | null
) =>
  invoke<ConversationDetail>("history_get_conversation_page", {
    conversationId,
    beforeSeq: beforeSeq ?? null,
    limit: limit ?? null,
  });

export const historyNewConversation = () =>
  invoke<ConversationSummary>("history_new_conversation");

export const historySetActiveConversation = (conversationId: string) =>
  invoke<void>("history_set_active_conversation", { conversationId });

export const historyMarkSeen = (conversationId: string) =>
  invoke<void>("history_mark_seen", { conversationId });

export const historyClearConversation = (conversationId: string) =>
  invoke<void>("history_clear_conversation", { conversationId });

export const historyDeleteConversation = (conversationId: string) =>
  invoke<HistoryBootstrap>("history_delete_conversation", { conversationId });

export const historyForkConversation = (
  conversationId: string,
  uptoSeq?: number | null
) => invoke<ConversationSummary>("history_fork_conversation", { conversationId, uptoSeq });

export const historyRenameConversation = (conversationId: string, title: string) =>
  invoke<void>("history_rename_conversation", { conversationId, title });

File: src-tauri/src/services/history.rs
//! Tauri command surface for conversation history.
//!
//! Storage is implemented by `crate::plugins::history::HistoryStore`.

use crate::plugins::history::HistoryStore;
pub use crate::plugins::history::{
    ConversationDetail, ConversationMessage, ConversationSummary, HistoryBootstrap, HistoryError,
};

#[tauri::command]
pub async fn history_bootstrap(
    store: tauri::State<'_, HistoryStore>,
) -> Result<HistoryBootstrap, HistoryError> {
    store.bootstrap().await
}

#[tauri::command]
pub async fn history_list_conversations(
    store: tauri::State<'_, HistoryStore>,
) -> Result<Vec<ConversationSummary>, HistoryError> {
    store.list_conversations().await
}

#[tauri::command]
pub async fn history_get_conversation(
    store: tauri::State<'_, HistoryStore>,
    conversation_id: String,
) -> Result<ConversationDetail, HistoryError> {
    store.get_conversation(&conversation_id).await
}

#[tauri::command]
pub async fn history_get_conversation_page(
    store: tauri::State<'_, HistoryStore>,
    conversation_id: String,
    before_seq: Option<u32>,
    limit: Option<u32>,
) -> Result<ConversationDetail, HistoryError> {
    store
        .get_conversation_page(&conversation_id, before_seq, limit)
        .await
}

#[tauri::command]
pub async fn history_new_conversation(
    store: tauri::State<'_, HistoryStore>,
) -> Result<ConversationSummary, HistoryError> {
    store.create_conversation(None, true).await
}

#[tauri::command]
pub async fn history_set_active_conversation(
    store: tauri::State<'_, HistoryStore>,
    conversation_id: String,
) -> Result<(), HistoryError> {
    store.set_active_conversation_id(&conversation_id).await
}

#[tauri::command]
pub async fn history_mark_seen(
    store: tauri::State<'_, HistoryStore>,
    conversation_id: String,
) -> Result<(), HistoryError> {
    store.mark_seen(&conversation_id).await
}

#[tauri::command]
pub async fn history_clear_conversation(
    store: tauri::State<'_, HistoryStore>,
    conversation_id: String,
) -> Result<(), HistoryError> {
    store.clear_messages(&conversation_id).await
}

#[tauri::command]
pub async fn history_delete_conversation(
    store: tauri::State<'_, HistoryStore>,
    conversation_id: String,
) -> Result<HistoryBootstrap, HistoryError> {
    store.delete_conversation(&conversation_id).await
}

#[tauri::command]
pub async fn history_fork_conversation(
    store: tauri::State<'_, HistoryStore>,
    conversation_id: String,
    upto_seq: Option<u32>,
) -> Result<ConversationSummary, HistoryError> {
    store
        .fork_conversation(&conversation_id, upto_seq, true)
        .await
}

#[tauri::command]
pub async fn history_rename_conversation(
    store: tauri::State<'_, HistoryStore>,
    conversation_id: String,
    title: String,
) -> Result<(), HistoryError> {
    store.rename_conversation(&conversation_id, &title).await
}
